tests/Console/E2ETest.php

```diff
@@ -0,0 +1,243 @@
+    }
+
+    /**
+     * @group e2e
+     */
+    public function test_it_runs_configure_command_if_no_configuration()
+    {
+        chdir('tests/Fixtures/e2e/Unconfigured/');
+
+        $output = $this->runInfection(self::EXPECT_ERROR);
+
+        $this->assertContains(ConfigureCommand::NONINTERACTIVE_MODE_ERROR, $output);
+    }
+
+    /**
+     * @dataProvider e2eTestSuiteDataProvider
+     * @group e2e
+     */
+    public function test_it_runs_an_e2e_test_with_success(string $e2eTestName, string $fullPath)
+    {
+        $this->runOnE2EFixture($fullPath);
+    }
+
+    public function e2eTestSuiteDataProvider(): \Generator
+    {
+        $directories = Finder::create()
+            ->depth('== 0')
+            ->in(__DIR__ . '/../Fixtures/e2e/')
+            ->directories();
+
+        foreach ($directories as $dirName) {
+            if (file_exists($dirName . '/run_tests.bash')) {
+                // skipping non-standard tests
+                // specifically Memory_Limit - it is very slow to fail
+                continue;
+            }
+
+            yield [
+                // Makes PHPUnit report readable errors on failure:
+                // ...with data set #0 ('Foo_bar', ...
+                basename((string) $dirName),
+                $dirName,
+            ];
+        }
+    }
+
+    protected function tearDown()
+    {
+        if ($this->previousLoader) {
+            $this->previousLoader->unregister();
+        }
+
+        chdir($this->cwd);
+    }
+
+    private function runOnE2EFixture($path): string
+    {
+        $this->assertTrue(is_dir($path));
+        chdir($path);
+
+        $this->installComposerDeps();
+        $output = $this->runInfection(self::EXPECT_SUCCESS);
+
+        $this->assertRegExp('/You are running Infection with \w+ enabled./', $output);
+        $this->assertRegExp('/\d+ mutations were generated/', $output);
+        $this->assertRegExp('/\d+ mutants were killed/', $output);
+
+        if (isset($_SERVER['GOLDEN'])) {
+            copy('infection-log.txt', 'expected-output.txt');
+            $this->markTestSkipped('Saved golden output');
+        }
+
+        $this->assertFileEquals('expected-output.txt', 'infection-log.txt', sprintf('%s/expected-output.txt is not same as infection-log.txt (if that is OK, run GOLDEN=1 vendor/bin/phpunit)', getcwd()));
+
+        return $output;
+    }
+
+    private function installComposerDeps()
+    {
+        if (!file_exists('composer.json')) {
+            // Tests may have no deps to install
+            return;
+        }
+
+        if (!file_exists('vendor/autoload.php')) {
+            // Install deps only if there's none
+            $this->assertNotEmpty(getenv('PATH') ?: getenv('Path'), 'E2E tests need a system composer installed, but it could not be found without a PATH set');
+
+            $process = new Process(sprintf('%s %s', (new ComposerExecutableFinder())->find(), 'install'));
```

Comment:
Probably, it makes sense to cache the result of `(new ComposerExecutableFinder())->find()` between test cases as it does not change

Reply:
We can do it, but this part is usually done by `./tests/e2e_tests`. In most cases we never have to install any dependencies right from the test, but failing a test because of missing deps will be silly. In other words, we almost never have to run `(new ComposerExecutableFinder())->find()` from here.

Do you still think we need to optimize it here?

Reply:
no, thanks